File: 04_sorting/04_07_insertion_sort_recursive/insertion_sort_recursive.cpp
#include <bits/stdc++.h>

using namespace std;

void printArray(int arr[], int n)
{
    for (size_t i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
}

void insertionSortRecursive(int arr[], int n, int length)
{
    if (length <= n)
    {
        return;
    }

    for (size_t j = n; j > 0; j--)
    {
        if (arr[j] < arr[j - 1])
        {
            swap(arr[j], arr[j - 1]);
        }
    }

    insertionSortRecursive(arr, n + 1, length);
}

int main()
{
    int arr[] = {13, 46, 24, 52, 20, 9};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Before:-" << endl;
    printArray(arr, n);
    insertionSortRecursive(arr, 0, n);
    cout << endl;
    cout << "After:-" << endl;
    printArray(arr, n);
}

File: 04_sorting/04_07_insertion_sort_recursive/InsertionSortRecursive.java
public class InsertionSortRecursive {
    public static void main(String[] args) {
        int arr[] = { 13, 46, 24, 52, 20, 9 };
        System.out.println("Before:-");
        print(arr);
        insertionSortRecursive(arr, 0);
        System.out.println();
        System.out.println("After:-");
        print(arr);
    }

    private static void insertionSortRecursive(int arr[], int n) {
        if (n >= arr.length) {
            return;
        }
        for (int j = n; j > 0; j--) {
            if (arr[j] < arr[j - 1]) {
                swap(arr, j, j - 1);
            }
        }
        insertionSortRecursive(arr, n + 1);
    }

    private static void swap(int[] arr, int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    private static void print(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}
